core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java

```diff
@@ -0,0 +1,478 @@
+                @Override
+                public RecoveryTarget apply(IndexShard indexShard, DiscoveryNode node) {
+                    return new RecoveryTarget(indexShard, node, recoveryListener) {
+                        @Override
+                        public void renameAllTempFiles() throws IOException {
+                            super.renameAllTempFiles();
+                            recoveryBlocked.countDown();
+                            try {
+                                releaseRecovery.await();
+                            } catch (InterruptedException e) {
+                                throw new IOException("terminated by interrupt", e);
+                            }
+                        }
+                    };
+                }
+            });
+
+            recoveryBlocked.await();
+            docs += shards.indexDocs(randomInt(20));
+            releaseRecovery.countDown();
+            recoveryFuture.get();
+
+            shards.assertAllEqual(docs);
+        }
+    }
+
+
+    @Before
+    public void setup() {
+        threadPool = new ThreadPool(
+            Settings.builder()
+                .put("node.name", getClass().getName())
+                .build()
+        );
+    }
+
+    @After
+    public void destroy() {
+        threadPool.shutdown();
+    }
+
+    private Store createStore(IndexSettings indexSettings, ShardPath shardPath) throws IOException {
+        final ShardId shardId = shardPath.getShardId();
+        final DirectoryService directoryService = new DirectoryService(shardId, indexSettings) {
+            @Override
+            public Directory newDirectory() throws IOException {
+                return newFSDirectory(shardPath.resolveIndex());
+            }
+
+            @Override
+            public long throttleTimeInNanos() {
+                return 0;
+            }
+        };
+        return new Store(shardId, indexSettings, directoryService, new DummyShardLock(shardId));
+    }
+
+    protected ReplicationGroup createGroup(int replicas) throws IOException {
+        final Path homePath = createTempDir();
+        Settings build = Settings.builder().put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, replicas)
+            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)
+            .build();
+        IndexMetaData metaData = IndexMetaData.builder(index.getName()).settings(build).primaryTerm(0, 1).build();
+        return new ReplicationGroup(metaData, homePath);
+    }
+
+    private DiscoveryNode getDiscoveryNode(String id) {
+        return new DiscoveryNode(id, id, DummyTransportAddress.INSTANCE, Collections.emptyMap(),
+            Collections.singleton(DiscoveryNode.Role.DATA), Version.CURRENT);
+    }
+
+    private IndexShard newShard(boolean primary, DiscoveryNode node, IndexMetaData indexMetaData, Path homePath) throws IOException {
```

Comment:
I think that is probably where it will end up being, yeah. Currently it makes a little too many assumptions on the environment it runs it for me to feel comfortably with moving it - it creates a new path for the shard and uses fixed mappings. I suggest we keep here for now and promote it once what the api should be crystallizes? if you have a concrete suggestion I'll be happy to use it and promote the method.